### ISSUE TITLE
Consider removing `mutex_atomic`

### ISSUE THREAD

**Issue by carllerche** [rust-lang/rust-clippy#4295]
There are many reasons to use mutexes over atomics. Clippy is unable to correctly determine if an atomic is sufficient to replace the mutex use. Attempting to lint over could lead to pushing users to introduce bugs in their code.

I recommend removing this lint.

**Comment by Manishearth** [comment]
I feel like this lint is correct most of the time; and clippy lints are supposed to be used with a liberal sprinkling of `allow()`. There needs to be stronger, more detailed justification for removing a lint.

**Comment by carllerche** [comment]
My main point is the potential harm of taking action based on a false positive is high compared to other provided lints.

**Comment by Manishearth** [comment]
I disagree, lints aren't as strong signals as compiler warnings, and Clippy's warnings are meant to be taken with a grain of salt. Here the signal is "there might be an issue", if you determine it isn't an issue, `allow()` it.

If you really think there's a problem here a stronger case needs to be made, with lots of examples.

**Comment by Manishearth** [comment]
> I have only experienced a 100% false positive rate

Like I said, clippy is supposed to be used with a liberal sprinkling of `allow()`. Some lints just don't make sense for some codebases. It's not possible for our lints to be totally globally applicable. It's up to the codebase to make this judgement.

**Comment by carllerche** [comment]
Ok, feel free to close if you disagree.

**Comment by Manishearth** [comment]
(leaving open since other maintainers, and other people, may have different opinions)

**Comment by flip1995** [comment]
The lint mentions in its `Known problems` section that
```
This lint cannot detect if the mutex is actually used for waiting before a critical section.
```

I'm with @Manishearth here. I think in the general case this lint is correct and `Atomic`s are sufficient. If you really need the mutex you should just allow the lint.

Removing this lint would require to show (with examples), that you need a Mutex in the general case.

**Comment by thombles** [comment]
I want to add a +1 since I ran into this reviewing a PR the other day where the submitter defaulted to using `Relaxed`. Replacing `Mutex<bool>` with `AtomicBool` like it shows [in the example](https://rust-lang.github.io/rust-clippy/master/index.html#mutex_atomic) is easy, but now you need to reason correctly about `Ordering`. Speaking only for myself, I've read about three articles about memory ordering and I still don't know which combination of `Acquire` and `Release` to use in a given situation. I happen to know that using `SeqCst` everywhere is safe, maybe because I read that on stack overflow once. That's a whole lot of homework that goes unmentioned and if you don't follow through on it you might introduce heisenbugs.

I'm quite happy to use the allow feature as noted above but I feel like the lint is a footgun to begin with. Those who know how to use atomics correctly will notice this potential performance issue the moment they see it.

**Comment by taiki-e** [comment]
Filed #8260 to downgrade this lint to `nursery` group that allowed by default.

Seems that Carl originally proposed to remove this lint completely, but the maintainers rejected it, so downgrading is a reasonable compromise for now.

**Comment by Nilstrieb** [comment]
I agree that this lint is bad and should be removed. Atomics are very complicated and very easy to get wrong. Mutexes on the other hand are simple and a lot harder to get wrong. People who don't think of using an atomic in the first place shouldn't be directed to using an atomic instead, the mutex will be fine.

This like feels a little like if clippy suggested replacing safe but a little more expensive type casts with transmute. Yes, it's faster and often works just as well, but it has a higher maintenance burden and more risk of getting it wrong.

We want people to write correct code, not the fastest code imaginable. And if someone knows atomics well enough to use them, they will probably not need this lint in the first place.

**Comment by rdrpenguin04** [comment]
It seems this is universally agreed upon that this lint is bad. Should someone make a PR to move this to `restriction` or something like that?

**Comment by llogiq** [comment]
I'm actually not sure if `restriction` is the right group here. Maybe put it into nursery instead and also add a comment to sum up this discussion to the lint docs.

**Comment by taiki-e** [comment]
Well, this lint is already in nursery since https://github.com/rust-lang/rust-clippy/pull/8260.

**Comment by rdrpenguin04** [comment]
The reason I feel `nursery` isn't a good fit is because, in general, the lints in `nursery` at least have potential to do the correct thing if the lint is finished; it's a "nursery" for future lints. In this case, the lint seems to be unhelpful in general, and the only change I see of it being useful is as a restriction. My own opinion though.

**Comment by ruuda** [comment]
This lint is also counter-productive advice when the `Mutex<bool>` is used with [`std::sync::Condvar::wait`](https://doc.rust-lang.org/std/sync/struct.Condvar.html#method.wait) and variations, because `Condvar::wait` needs the mutex guard and can’t work with an `AtomicBool`. In fact, the example in the `Condvar` documentation for its canonical usage relies on `Mutex<bool>`. This was already raised separately in #1516, but I thought I’d highlight it here, because the lint’s docs link to this issue in particular.
